File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/android/chain_checks.py
"""
Per-phase certificate-chain validators for Android Key Attestation.
"""

from __future__ import annotations

import json
import logging
import re
from datetime import datetime, timezone
from typing import Optional

from cryptography import x509

from ..logging import track_event, track_exception
from ..base64_utils import decode_base64
from ..cert_utils import (
    GOOGLE_HARDWARE_ATTESTATION_ROOT_CAS,
    matches_pinned_ca,
    pem_to_der,
    validate_certificate_path,
)
from .keymaster_ext import (
    extract_attestation_challenge_from_cert,
    is_hardware_attestation_security_level,
    parse_key_description,
)
from .revocation import check_certificate_revocation, fetch_revocation_status_list
from .types import AndroidPhaseFail

log = logging.getLogger("android.chain_checks")


def parse_android_attest_json(attest_json: str, base_props: dict) -> dict:
    try:
        data = json.loads(attest_json)
    except Exception as err:
        log.error("[AndroidAuth] Failed to parse attestJson")
        track_exception(err, {"source": "verify_android_auth.parseAttestJson", **base_props})
        raise AndroidPhaseFail("ATTEST_JSON_PARSE_ERROR", "Invalid attestation: failed to parse attestJson") from err

    cc = data.get("certificateChain")
    if cc is None or not isinstance(cc, list):
        raise AndroidPhaseFail("INVALID_CHAIN_FORMAT", "Invalid attestation: missing or invalid certificateChain")
    if len(cc) == 0:
        raise AndroidPhaseFail("EMPTY_CHAIN", "Invalid attestation: empty certificateChain")
    return data


def build_android_chain_der_buffers(attest_data: dict, public_cert_pem: Optional[str]) -> dict:
    if not public_cert_pem:
        raise AndroidPhaseFail("MISSING_PUBLIC_CERT", "Missing publicCert in message data")
    cert_chain_der = [decode_base64(c) for c in attest_data["certificateChain"]]
    leaf_cert_der = pem_to_der(public_cert_pem)
    return {"certChainDer": cert_chain_der, "leafCertDer": leaf_cert_der}


def verify_leaf_bound_to_chain0(leaf_cert_der: bytes, cert_chain_der: list[bytes]) -> None:
    if leaf_cert_der != cert_chain_der[0]:
        raise AndroidPhaseFail(
            "LEAF_CHAIN_MISMATCH",
            "Certificate chain verification failed: publicCert must match certificateChain[0]",
            extras={"chainLength": len(cert_chain_der)},
            extra_props={"chainLength": len(cert_chain_der)},
        )


def validate_android_chain_validity_dates(
    leaf_cert_der: bytes,
    cert_chain_der: list[bytes],
    base_props: dict,
) -> dict:
    try:
        leaf_cert = x509.load_der_x509_certificate(leaf_cert_der)
        now = datetime.now(timezone.utc)
        leaf_warning = None
        if now < leaf_cert.not_valid_before_utc or now > leaf_cert.not_valid_after_utc:
            leaf_warning = (
                f"Leaf certificate validity warning: valid from "
                f"{leaf_cert.not_valid_before_utc.isoformat()} to {leaf_cert.not_valid_after_utc.isoformat()}"
            )
            track_event(
                "AndroidAuth.LeafCertValidityWarning",
                {**base_props,
                 "validFrom": leaf_cert.not_valid_before_utc.isoformat(),
                 "validTo": leaf_cert.not_valid_after_utc.isoformat()},
            )

        for i, der in enumerate(cert_chain_der):
            c = x509.load_der_x509_certificate(der)
            if now < c.not_valid_before_utc or now > c.not_valid_after_utc:
                raise AndroidPhaseFail(
                    "CHAIN_CERT_NOT_VALID",
                    f"Certificate at index {i} is not valid (valid from {c.not_valid_before_utc.isoformat()} to {c.not_valid_after_utc.isoformat()})",
                    extras={"chainLength": len(cert_chain_der) + 1},
                    extra_props={
                        "chainIndex": i,
                        "certValidFrom": c.not_valid_before_utc.isoformat(),
                        "certValidTo": c.not_valid_after_utc.isoformat(),
                    },
                )
        return {"leafCertValidityWarning": leaf_warning}
    except AndroidPhaseFail:
        raise
    except Exception as err:
        track_exception(err, {"source": "verify_android_auth.validityDates", **base_props})
        raise AndroidPhaseFail("CERT_VALIDITY_ERROR", "Failed to validate certificate validity dates") from err


def verify_android_chain_signatures_and_root(cert_chain_der: list[bytes]) -> dict:
    root = cert_chain_der[-1]
    if not matches_pinned_ca(root, GOOGLE_HARDWARE_ATTESTATION_ROOT_CAS):
        raise AndroidPhaseFail(
            "ROOT_CA_MISMATCH",
            "Certificate chain verification failed: root CA does not match any pinned CA",
            extras={"chainLength": len(cert_chain_der)},
            extra_props={"chainLength": len(cert_chain_der)},
        )

    if not validate_certificate_path(cert_chain_der, GOOGLE_HARDWARE_ATTESTATION_ROOT_CAS):
        raise AndroidPhaseFail(
            "CHAIN_PATH_INVALID",
            "Certificate path validation failed",
            extras={"chainLength": len(cert_chain_der)},
            extra_props={"chainLength": len(cert_chain_der)},
        )

    root_ca_subject = "unknown"
    try:
        root_ca_subject = x509.load_der_x509_certificate(root).subject.rfc4514_string()
    except Exception as err:
        log.error("Failed to parse root CA subject: %s", err)
    return {"rootCASubject": root_ca_subject}


def verify_attestation_security_level(chain0_der: bytes) -> None:
    description = parse_key_description(chain0_der)
    if description is None:
        raise AndroidPhaseFail(
            "KEYMASTER_EXT_MISSING",
            "Android Key Attestation extension (OID 1.3.6.1.4.1.11129.2.1.17) missing or unparseable on leaf attestation cert",
        )
    if not is_hardware_attestation_security_level(description):
        raise AndroidPhaseFail(
            "ATTESTATION_SECURITY_LEVEL_INVALID",
            "Android Key Attestation verification failed: attestationSecurityLevel must be TrustedEnvironment or StrongBox",
            extra_props={"attestationSecurityLevel": description.attestationSecurityLevel},
        )


def verify_android_chain_revocations(leaf_cert_der: bytes, cert_chain_der: list[bytes]) -> None:
    status_list = fetch_revocation_status_list()

    leaf_check = check_certificate_revocation(leaf_cert_der, status_list)
    if leaf_check.get("isRevoked"):
        raise AndroidPhaseFail(
            "LEAF_REVOKED",
            f"Leaf certificate has been {leaf_check.get('status')}: {leaf_check.get('reason')}",
            extras={"chainLength": len(cert_chain_der) + 1},
            extra_props={
                "revocationStatus": leaf_check.get("status"),
                "revocationReason": leaf_check.get("reason"),
            },
        )

    for i, der in enumerate(cert_chain_der):
        rc = check_certificate_revocation(der, status_list)
        if rc.get("isRevoked"):
            raise AndroidPhaseFail(
                "CHAIN_CERT_REVOKED",
                f"Certificate at index {i} has been {rc.get('status')}: {rc.get('reason')}",
                extras={"chainLength": len(cert_chain_der) + 1},
                extra_props={
                    "chainIndex": i,
                    "revocationStatus": rc.get("status"),
                    "revocationReason": rc.get("reason"),
                },
            )


def verify_chain_binds_session_challenge(chain0_der: bytes, session_challenge_hash: Optional[str]) -> dict:
    if not session_challenge_hash:
        raise AndroidPhaseFail("MISSING_CHALLENGE_HASH", "Missing challengeHash in message data")
    if not re.fullmatch(r"[0-9a-f]+", session_challenge_hash, flags=re.IGNORECASE) or len(session_challenge_hash) % 2 != 0:
        raise AndroidPhaseFail("CHALLENGE_HASH_FORMAT", "challengeHash must be a hex string of even length")
    session_bytes = bytes.fromhex(session_challenge_hash)
    chain_challenge = extract_attestation_challenge_from_cert(chain0_der)
    if chain_challenge is None:
        raise AndroidPhaseFail(
            "KEYMASTER_EXT_MISSING",
            "Android Key Attestation extension (OID 1.3.6.1.4.1.11129.2.1.17) missing or unparseable on leaf attestation cert",
        )
    if chain_challenge != session_bytes:
        raise AndroidPhaseFail(
            "CHAIN_CHALLENGE_MISMATCH",
            "Android Key Attestation chain challenge does not match session challengeHash",
            extra_props={
                "expectedPrefix": session_bytes.hex()[:12],
                "gotPrefix": chain_challenge.hex()[:12],
            },
        )
    return {"chainChallengeHex": chain_challenge.hex()}

File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/android/chain_checks.ts
/**
 * Per-phase certificate-chain validators for Android Key Attestation.
 *
 *   parseAndroidAttestJson              — JSON.parse + structural checks
 *   buildAndroidChainDerBuffers         — base64 chain + PEM leaf → DER buffers
 *   verifyLeafBoundToChain0             — auth cert must be the attested chain leaf
 *   validateAndroidChainValidityDates   — notBefore/notAfter sweep
 *   verifyAndroidChainSignaturesAndRoot — cert[i] signed by cert[i+1] + pinned root
 *   verifyAttestationSecurityLevel      — require TEE or StrongBox attestation
 *   verifyAndroidChainRevocations       — hits the Google revocation list
 *   verifyChainBindsSessionChallenge    — keymaster ext == sessionChallengeHash
 *
 * Each helper returns either its result payload or an `AndroidPhaseFail` that
 * the main verifier turns into the matching `failVerify` call.
 */
import crypto from 'crypto';
import { pemToDer, validateCertificatePath, matchesPinnedCA } from '../cert_utils';
import { GOOGLE_HARDWARE_ATTESTATION_ROOT_CAS } from './google_roots';
import { trackEvent, trackException } from '../logging';
import { androidAttestationSchema, AndroidPhaseFail, type AndroidAttestationJson, type BaseProps } from './types';
import { fetchRevocationStatusList, checkCertificateRevocation } from './revocation';
import {
  extractAttestationChallengeFromCert,
  isHardwareAttestationSecurityLevel,
  parseKeyDescription,
} from './keymaster_ext';

/**
 * Phase 1: parse the JSON envelope and validate the certificateChain field
 * exists, is an array, and is non-empty.
 */
export function parseAndroidAttestJson(
  attestJson: string,
  baseProps: BaseProps,
): AndroidAttestationJson | AndroidPhaseFail {
  let parsed: unknown;
  try {
    parsed = JSON.parse(attestJson);
  } catch (parseErr) {
    trackException(parseErr, { source: 'verifyAndroidAuth.parseAttestJson', ...baseProps });
    return new AndroidPhaseFail(
      'ATTEST_JSON_PARSE_ERROR',
      'Invalid attestation: failed to parse attestJson',
    );
  }

  const result = androidAttestationSchema.safeParse(parsed);
  if (!result.success) {
    const issue = result.error.issues[0];
    if (issue.path[0] === 'token') {
      return new AndroidPhaseFail('MISSING_INTEGRITY_TOKEN', 'Play Integrity token must be a string');
    }
    if (issue.path[0] === 'certificateChain' && issue.code === 'too_small') {
      return new AndroidPhaseFail('EMPTY_CHAIN', 'Invalid attestation: empty certificateChain');
    }
    return new AndroidPhaseFail(
      'INVALID_CHAIN_FORMAT',
      'Invalid attestation: missing or invalid certificateChain',
    );
  }

  trackEvent('AndroidAuth.AttestationParsed', {
    hasIntegrityToken: !!result.data.token,
    integrityTokenLength: result.data.token?.length ?? 0,
    hasCertificateChain: true,
    certificateChainLength: result.data.certificateChain.length,
  });
  return result.data;
}

/**
 * Phase 2: turn the base64 chain + PEM leaf into DER buffers.
 */
export function buildAndroidChainDerBuffers(
  attestData: AndroidAttestationJson,
  publicCertPem: string | undefined,
): { certChainDer: Buffer[]; leafCertDer: Buffer } | AndroidPhaseFail {
  if (!publicCertPem) {
    return new AndroidPhaseFail('MISSING_PUBLIC_CERT', 'Missing publicCert in message data');
  }

  const certChainDer: Buffer[] = attestData.certificateChain.map((base64Cert) =>
    Buffer.from(base64Cert, 'base64'),
  );
  const leafCertDer = pemToDer(publicCertPem);

  trackEvent('AndroidAuth.CertificateChainDecoded', {
    certificateChainLength: certChainDer.length,
    leafCertificateLength: leafCertDer.length,
    chainCertificateLengths: certChainDer.map(cert => cert.length),
  });
  return { certChainDer, leafCertDer };
}

/**
 * Phase 3: the auth cert (messageData.publicCert) must be byte-identical to
 * certificateChain[0], the hardware-attested leaf. Accepting a separate cert
 * merely signed by chain[0] would allow the hardware key to delegate trust to
 * an exportable software key.
 */
export function verifyLeafBoundToChain0(
  leafCertDer: Buffer,
  certChainDer: Buffer[],
): null | AndroidPhaseFail {
  const leafMatchesChain0 = leafCertDer.equals(certChainDer[0]);
  if (leafMatchesChain0) return null;

  try {
    const leafCert = new crypto.X509Certificate(leafCertDer);
    const chain0Cert = new crypto.X509Certificate(certChainDer[0]);
    trackEvent('AndroidAuth.LeafChainMismatch', {
      leafSubject: leafCert.subject,
      leafIssuer: leafCert.issuer,
      leafFingerprintPrefix: leafCert.fingerprint256.slice(0, 23),
      chain0Subject: chain0Cert.subject,
      chain0Issuer: chain0Cert.issuer,
      chain0FingerprintPrefix: chain0Cert.fingerprint256.slice(0, 23),
      certificateChainLength: certChainDer.length,
    });
  } catch (error) {
    trackException(error, { source: 'verifyAndroidAuth.compareLeafAndChain' });
  }
  return new AndroidPhaseFail(
    'LEAF_CHAIN_MISMATCH',
    'Certificate chain verification failed: publicCert must match certificateChain[0]',
    { chainLength: certChainDer.length },
    { chainLength: certChainDer.length },
  );
}

/**
 * Phase 4: validity-date sweep across leaf + chain. Leaf-out-of-validity is
 * logged as a warning and surfaced to the caller; any chain cert out of
 * validity hard-fails, even under DEBUG_MODE.
 */
export function validateAndroidChainValidityDates(
  leafCertDer: Buffer,
  certChainDer: Buffer[],
  baseProps: BaseProps,
): { leafCertValidityWarning?: string } | AndroidPhaseFail {
  try {
    const leafCert = new crypto.X509Certificate(leafCertDer);
    const now = new Date();
    const validFrom = new Date(leafCert.validFrom);
    const validTo = new Date(leafCert.validTo);

    let leafCertValidityWarning: string | undefined;
    if (now < validFrom || now > validTo) {
      leafCertValidityWarning = `Leaf certificate validity warning: valid from ${leafCert.validFrom} to ${leafCert.validTo}`;
      trackEvent('AndroidAuth.LeafCertValidityWarning', {
        ...baseProps,
        validFrom: leafCert.validFrom,
        validTo: leafCert.validTo,
      });
    }

    for (let i = 0; i < certChainDer.length; i++) {
      const cert = new crypto.X509Certificate(certChainDer[i]);
      const certValidFrom = new Date(cert.validFrom);
      const certValidTo = new Date(cert.validTo);

      if (now < certValidFrom || now > certValidTo) {
        return new AndroidPhaseFail(
          'CHAIN_CERT_NOT_VALID',
          `Certificate at index ${i} is not valid (valid from ${cert.validFrom} to ${cert.validTo})`,
          { chainLength: certChainDer.length + 1 },
          {
            chainIndex: i,
            certValidFrom: cert.validFrom,
            certValidTo: cert.validTo,
          },
        );
      }
    }

    return { leafCertValidityWarning };
  } catch (error) {
    trackException(error, { source: 'verifyAndroidAuth.validityDates', ...baseProps });
    return new AndroidPhaseFail(
      'CERT_VALIDITY_ERROR',
      'Failed to validate certificate validity dates',
    );
  }
}

/**
 * Phase 5: cert[i] must be signed by cert[i+1] for the entire chain, and the
 * root (last cert) must match one of the pinned Google Hardware Attestation
 * CAs.
 *
 * Returns the root CA subject string (best-effort, "unknown" if it doesn't
 * parse — purely for logging).
 */
export async function verifyAndroidChainSignaturesAndRoot(
  certChainDer: Buffer[],
): Promise<{ rootCASubject: string } | AndroidPhaseFail> {
  // pinned root CA
  const rootCert = certChainDer[certChainDer.length - 1];
  if (!matchesPinnedCA(rootCert, GOOGLE_HARDWARE_ATTESTATION_ROOT_CAS)) {
    try {
      const parsedRoot = new crypto.X509Certificate(rootCert);
      trackEvent('AndroidAuth.RootCAMismatch', {
        rootSubject: parsedRoot.subject,
        rootIssuer: parsedRoot.issuer,
        rootFingerprintPrefix: parsedRoot.fingerprint256.slice(0, 23),
        certificateChainLength: certChainDer.length,
        pinnedRootCount: GOOGLE_HARDWARE_ATTESTATION_ROOT_CAS.length,
      });
    } catch (error) {
      trackException(error, { source: 'verifyAndroidAuth.parseUntrustedRoot' });
    }
    return new AndroidPhaseFail(
      'ROOT_CA_MISMATCH',
      'Certificate chain verification failed: root CA does not match any pinned CA',
      { chainLength: certChainDer.length },
      { chainLength: certChainDer.length },
    );
  }

  if (!await validateCertificatePath(certChainDer, GOOGLE_HARDWARE_ATTESTATION_ROOT_CAS)) {
    return new AndroidPhaseFail('CHAIN_PATH_INVALID', 'Certificate path validation failed',
      { chainLength: certChainDer.length });
  }
  let rootCASubject = 'unknown';
  try {
    rootCASubject = new crypto.X509Certificate(rootCert).subject;
  } catch (error) {
    trackException(error, { source: 'verifyAndroidAuth.rootSubject' });
  }
  return { rootCASubject };
}

/**
 * Phase 6: require hardware-backed attestation from TEE or StrongBox.
 */
export function verifyAttestationSecurityLevel(
  chain0Der: Buffer,
): null | AndroidPhaseFail {
  const description = parseKeyDescription(chain0Der);
  if (!description) {
    return new AndroidPhaseFail(
      'KEYMASTER_EXT_MISSING',
      'Android Key Attestation extension (OID 1.3.6.1.4.1.11129.2.1.17) missing or unparseable on leaf attestation cert',
    );
  }
  if (!isHardwareAttestationSecurityLevel(description)) {
    return new AndroidPhaseFail(
      'ATTESTATION_SECURITY_LEVEL_INVALID',
      'Android Key Attestation verification failed: attestationSecurityLevel must be TrustedEnvironment or StrongBox',
      undefined,
      { attestationSecurityLevel: description.attestationSecurityLevel },
    );
  }
  return null;
}

/**
 * Phase 7: hit the Google revocation list and reject if either the leaf or
 * any chain cert is revoked/suspended. fetchRevocationStatusList fails closed
 * inside checkCertificateRevocation, so an unreachable list is also rejected.
 */
export async function verifyAndroidChainRevocations(
  leafCertDer: Buffer,
  certChainDer: Buffer[],
): Promise<null | AndroidPhaseFail> {
  const revocationStatusList = await fetchRevocationStatusList();

  const leafRevocation = checkCertificateRevocation(leafCertDer, revocationStatusList);
  if (leafRevocation.isRevoked) {
    return new AndroidPhaseFail(
      'LEAF_REVOKED',
      `Leaf certificate has been ${leafRevocation.status}: ${leafRevocation.reason}`,
      { chainLength: certChainDer.length + 1 },
      {
        revocationStatus: leafRevocation.status,
        revocationReason: leafRevocation.reason,
      },
    );
  }

  for (let i = 0; i < certChainDer.length; i++) {
    const revocationCheck = checkCertificateRevocation(certChainDer[i], revocationStatusList);
    if (revocationCheck.isRevoked) {
      return new AndroidPhaseFail(
        'CHAIN_CERT_REVOKED',
        `Certificate at index ${i} has been ${revocationCheck.status}: ${revocationCheck.reason}`,
        { chainLength: certChainDer.length + 1 },
        {
          chainIndex: i,
          revocationStatus: revocationCheck.status,
          revocationReason: revocationCheck.reason,
        },
      );
    }
  }
  return null;
}

/**
 * Phase 8: the keymaster attestation extension on the leaf must carry the
 * *current* session challengeHash. This is the primary session-binding proof
 * for Android — pre-rotation chains from earlier sessions are explicitly
 * rejected here, even when their certificate chain still validates
 * cryptographically.
 */
export function verifyChainBindsSessionChallenge(
  chain0Der: Buffer,
  sessionChallengeHash: string | undefined,
): { chainChallengeHex: string } | AndroidPhaseFail {
  if (!sessionChallengeHash) {
    return new AndroidPhaseFail(
      'MISSING_CHALLENGE_HASH',
      'Missing challengeHash in message data',
    );
  }
  if (!/^[0-9a-f]+$/i.test(sessionChallengeHash) || sessionChallengeHash.length % 2 !== 0) {
    return new AndroidPhaseFail(
      'CHALLENGE_HASH_FORMAT',
      'challengeHash must be a hex string of even length',
    );
  }
  const sessionChallengeBytes = Buffer.from(sessionChallengeHash, 'hex');
  const chainChallenge = extractAttestationChallengeFromCert(chain0Der);
  if (!chainChallenge) {
    return new AndroidPhaseFail(
      'KEYMASTER_EXT_MISSING',
      'Android Key Attestation extension (OID 1.3.6.1.4.1.11129.2.1.17) missing or unparseable on leaf attestation cert',
    );
  }
  if (!chainChallenge.equals(sessionChallengeBytes)) {
    return new AndroidPhaseFail(
      'CHAIN_CHALLENGE_MISMATCH',
      'Android Key Attestation chain challenge does not match session challengeHash',
      undefined,
      {
        expectedPrefix: sessionChallengeBytes.toString('hex').slice(0, 12),
        gotPrefix: chainChallenge.toString('hex').slice(0, 12),
      },
    );
  }
  return { chainChallengeHex: chainChallenge.toString('hex') };
}
